aten/src/ATen/core/type.cpp

```diff
@@ -702,6 +752,157 @@ std::string TupleType::python_str_impl(TypePrinter printer) const {
   return ss.str();
 }
 
+static std::vector<bool> findContiguous(
+    const at::IntArrayRef& sizes,
+    const at::IntArrayRef& strides) {
+  AT_ASSERT(sizes.size() == strides.size());
+  std::vector<bool> cont(sizes.size());
+  for (size_t i = 0; i < sizes.size(); ++i) {
+    const auto expected_stride =
+        (i + 1 < sizes.size()) ? sizes[i + 1] * strides[i + 1] : 1;
+    cont[i] = (strides[i] == expected_stride);
+  }
+  return cont;
+}
+
+std::tuple<std::vector<bool>, std::vector<size_t>> TensorType::
+    contiguityStrideIndices(at::IntArrayRef sizes, at::IntArrayRef strides) {
+  auto contiguity = findContiguous(sizes, strides);
```

Comment:
Shouldn't contiguity be determined after sorting? If we know that we are handling transposed data, for kernel generation, what we would care about is the contiguity information in ordered stride.

Reply:
correct, contiguity information is suppose to be from smallest to largest. So this probably needs to be done in verse order: first compute stride_indices, then contiguity form them.

Reply:
To make this problem even more convoluted, currently we sort dimensions by their stride and break tie on their semantic index (https://github.com/pytorch/pytorch/pull/36101/files/4366d3d892bd89811c11483474e0917f612c3f1d#diff-1eaac5e95929476c6522cd50bbb2432eR781)

However, it makes the contiguity information ambiguous.
If we have an unsqueezed tensor with size [4, 1, 8] and a stride [8, 1, 1], current implementation would compare contiguity between dimension [0/1] & [1/2] where it won't correctly identify that the meaningful dimension [0/2] are actually contiguous.

Size-1 dimension makes contiguity tricky to handle for two reasons: 1. stride could be wild; 2. it breaks the cumulative law of contiguity;
I think we should special case size-1 dimension to make our life easier:
1. We still need to preserve size-1 dimension when we sort the dimensions per their stride. (This is to ensure our memory_format stuff works properly)
2. size-1 tensor should transparently pass contiguity check and instead assess contiguity  information for closest neighboring dimensions with `size != 1`;

I'm in favor of Zach's idea of having a separate class to properly encapsulate this logic.